h2o-automl/src/main/java/ai/h2o/automl/targetencoding/TargetEncoder.java

```diff
@@ -0,0 +1,745 @@
+      }
+      else {
+        int foldColumnIndex = getColumnIndexByName(fr, foldColumnName);
+        groupByColumns = new int[]{teColumnIndex, foldColumnIndex};
+      }
+
+      AstGroup.AGG[] aggs = new AstGroup.AGG[2];
+
+      AstGroup.NAHandling na = AstGroup.NAHandling.ALL;
+      aggs[0] = new AstGroup.AGG(AstGroup.FCN.sum, targetIndex, na, (int) fr.vec(targetIndex).max() + 1);
+      aggs[1] = new AstGroup.AGG(AstGroup.FCN.nrow, targetIndex, na, (int) fr.vec(targetIndex).max() + 1);
+
+      Frame result = new AstGroup().performGroupingWithAggregations(fr, groupByColumns, aggs, -1).getFrame();
+      return FrameUtils.register(result);
+    }
+
+    Frame ensureTargetColumnIsNumericOrBinaryCategorical(Frame data, String targetColumnName) {
+        return ensureTargetColumnIsNumericOrBinaryCategorical(data, getColumnIndexByName(data, targetColumnName));
+    };
+
+    Frame ensureTargetColumnIsNumericOrBinaryCategorical(Frame data, int targetIndex) {
+        if (data.vec(targetIndex).isCategorical()){
+            Vec targetVec = data.vec(targetIndex);
+            if(targetVec.cardinality() == 2) {
+                return transformBinaryTargetColumn(data, targetIndex);
+            }
+            else {
+                throw new IllegalStateException("`target` must be a binary vector");
+            }
+        }
+        else {
+            if(! data.vec(targetIndex).isNumeric()) {
+                throw new IllegalStateException("`target` must be a numeric or binary vector");
+            }
+            return data;
+        }
+    };
+
+
+  public Map<String, Frame> prepareEncodingMap(Frame data, String targetColumnName, String foldColumnName) {
+      return prepareEncodingMap( data, targetColumnName, foldColumnName, false);
+  }
+
+    String[] getColumnNamesBy(Frame data, int[] columnIndexes) {
+        String [] allColumnNames = data._names.clone();
+        ArrayList<String> columnNames = new ArrayList<String>();
+
+        for(int idx : columnIndexes) {
+            columnNames.add(allColumnNames[idx]);
+        }
+        return columnNames.toArray(new String[columnIndexes.length]);
+    }
+
+    private Frame execRapidsAndGetFrame(String astTree) {
+        Val val = Rapids.exec(astTree);
+        Frame res = val.getFrame();
+        res._key = Key.make();
+        DKV.put(res);
+        return res;
+    }
+
+    //TODO We might want to introduce parameter that will change this behaviour. We can treat NA's as extra class.
+    Frame filterOutNAsFromTargetColumn(Frame data, int targetColumnIndex) {
+        return data.filterOutNAsInColumn(targetColumnIndex);
+    }
+
+    Frame imputeNAsForColumn(Frame data, String teColumnName, String strToImpute) {
+      int columnIndex = getColumnIndexByName(data, teColumnName);
+      Vec currentVec = data.vec(columnIndex);
+      int indexForNACategory = currentVec.cardinality(); // Warn: Cardinality returns int but it could be larger than it for big datasets
```

Comment:
the limit on number of categoricals is 10 million

Reply:
removed comment